Dedupe fixes (many)

eileenmcnaughton opened the pull request:
* [CRM-18681: Permissioning issues with batch merge job](https://issues.civicrm.org/jira/browse/CRM-18681)

eileenmcnaughton commented on CRM/Dedupe/Merger.php:
```diff
@@ -1812,10 +1812,9 @@ public static function moveAllBelongings($mainId, $otherId, $migrationInfo) {
       civicrm_api3('contact', 'delete', array('id' => $otherId));
-      CRM_Core_BAO_PrevNextCache::deleteItem($otherId);
     }
+    CRM_Core_BAO_PrevNextCache::deleteItem($otherId);
```
I'm feeling pretty uncomfortable with the permission check in this function as it seems that if they don't have permission to merge that should be checked earlier on. Regardless - if the belongings have been shifted we should pull the semi-deleted contact out of any sets it belongs to. If we leave it in the merge set we can create a code loop in the unit tests & we can reasonably assume it's fairly invalid in most searches now. However, I would argue we should never reach this point in that only contacts who can run a test should be able to get to the move all belongings fn

JoeMurray commented:
@JKingsnorth would you be interested in reviewing the PR?

JoeMurray commented:
@JKingsnorth I couldn't find a second dedupe PR from @eileenmcnaughton - can you help me understand if it is one from @nielosz ?

eileenmcnaughton commented:
This PR has lots of commits in it rather than there being 2 I believe :-)

eileenmcnaughton commented:
NB - I can split off it you want to start looking before I resolve the tests.... but I'll get onto them today I think

nielosz commented:
@JoeMurray, Maybe I can help, but I think don't fully understand your comment :/

JoeMurray commented:
@nielosz Thanks for your offer. @JKingsnorth had said he was willing to work on 2 PRs from @eileenmcnaughton to do with dupes, but I only saw one. I think @eileenmcnaughton is now suggesting that she may split the PR to make review (and eventual merging) simpler.

eileenmcnaughton commented:
@JKingsnorth are you going to be able to tear yourself away from voting tomorrow enough to be able to QA this? It does contain a tonne of fixes so it's a big job :-( but it does fix those criticals & some other things we IDd

JKingsnorth commented:
Hi there. I'll try to give it a quick look over today, I don't think I can
give it a line by line review though. Sorry to not be so involved this
cycle, everything happening in the same week here :)
On 23 Jun 2016 02:46, "Eileen McNaughton" <notifications@github.com> wrote:

> @JKingsnorth <https://github.com/JKingsnorth> are you going to be able to
> tear yourself away from voting tomorrow enough to be able to QA this? It
> does contain a tonne of fixes so it's a big job :-( but it does fix those
> criticals & some other things we IDd
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/civicrm/civicrm-core/pull/8456#issuecomment-227927911>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE-JuYYB8rlxzpnTBPm1XSm9cOczRJf4ks5qOeVxgaJpZM4InIAt>
> .
>

eileenmcnaughton commented:
Yeah I know it's hectic. I'm feeling pretty OK about it as the test 
coverage is much improved since last round.

Eileen

On 23/06/2016 6:24 p.m., John Kingsnorth wrote:
> Hi there. I'll try to give it a quick look over today, I don't think I can
> give it a line by line review though. Sorry to not be so involved this
> cycle, everything happening in the same week here :)
> On 23 Jun 2016 02:46, "Eileen McNaughton" <notifications@github.com> 
> wrote:
>
> > @JKingsnorth <https://github.com/JKingsnorth> are you going to be 
> able to
> > tear yourself away from voting tomorrow enough to be able to QA this? It
> > does contain a tonne of fixes so it's a big job :-( but it does fix 
> those
> > criticals & some other things we IDd
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > 
> <https://github.com/civicrm/civicrm-core/pull/8456#issuecomment-227927911>,
> > or mute the thread
> > 
> <https://github.com/notifications/unsubscribe/AE-JuYYB8rlxzpnTBPm1XSm9cOczRJf4ks5qOeVxgaJpZM4InIAt>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/civicrm/civicrm-core/pull/8456#issuecomment-227960885>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AAUhtDj8n-c3CaWIvDl68BK2QcVH47Ftks5qOiazgaJpZM4InIAt>.
>



---
This email has been checked for viruses by Avast antivirus software.
https://www.avast.com/antivirus

JKingsnorth commented on CRM/Contact/Form/DedupeFind.php:
```diff
@@ -51,6 +51,7 @@ public function buildQuickForm() {
     $groupList = array('' => ts('- All Contacts -')) + CRM_Core_PseudoConstant::nestedGroup();
 
     $this->add('select', 'group_id', ts('Select Group'), $groupList, FALSE, array('class' => 'crm-select2 huge'));
+    $this->add('text', 'limit', ts('No of contacts to find matches for '));
```
We should at least write out 'Number of contacts to find matches for'. Or possibly 'Maximum number of duplicates to display'?

eileenmcnaughton replied:
Good point - it's definitely not the latter though - it is as many matches are found for the first x number of contacts - so 1000 contacts might return 5 results or 5000.... Not sure the best wording for that.

JKingsnorth replied:
I haven't been able to get it to work anyway, limiting to 20 still shows 67 duplicates. I think it's a bit misleading to only search the first 1000 contacts... shouldn't it limit it to 1000 results instead?

JKingsnorth replied:
Ah it seems to be a caching thing. If you search for 1000 and then for 20 the results don't change. Pesky.

eileenmcnaughton replied:
The reason it does the number of contacts not results is that it gives you a way to control the size of the query. If you have a DB of 200,000 contacts it's impossible to know how many contacts you have to do a match.find on to get 1000 results - it could be 100 - it could be 20,000 - so you would need to maybe batch by 100 contacts or something until you get to 1000 results.

JKingsnorth replied:
OK, but I think that makes it more complicated from the UI point of view, and more confusing for end users. If you just increase the limit on the UI then you're searching the same people again and again, since you can't specify the range to search. I agree that limiting is useful, but I'm not sure how this solution actually improves anything in the UI since you'd have to select the first 1000 contacts in the DB to look for duplicates for. And then the first 2000, etc. etc. You still end up having to do the 'massive' query to search the entire contact base for duplicates.

At least I'd recommend having a 'no limit' option which is selected by default.

eileenmcnaughton replied:
the main reason for it is to protect people from crashing the db inadvertently - which is what no limit does on a large group or site - that's the problem I'm trying find a solution for

JKingsnorth commented:
The 'flip' issue doesn't seem to be fixed by this PR - is there another commit that's required? If you 'force' merge two records that have been flipped, the wrong one is deleted.

eileenmcnaughton commented:
Hmm I thought it was fixed :-( I certainly tackled the lack of flipping in the DB itself

JKingsnorth commented:
Confirmed the manual tests all pass, except for the 'flip' issue described above.

Confirmed this fixes:
CRM-18577 (postcode searches)
CRM-18547 (group filter persistence)

eileenmcnaughton commented:
OK - what is your thinking about what needs to happen pre-merge?

JKingsnorth commented:
* Would be good to get the 'flip' issue tested and working. It works when you 'flip' and then click to manually merge. But not when you 'flip', safe merge (resulting in conflicT) and then choose to force merge.
* Would be good to resolve the questions around the 'limit' field

Thanks for your work on this Eileen =]

JKingsnorth commented:
After a useful discussion on MM, we have this solution for the 'limit' UI:

o Check [ 1000 ] contacts, skipping the first [ 5 ] contacts
o Check the entire database (this might cause the server to crash, but I'm feelin' lucky)

JKingsnorth commented:
As a follow-up, we could just solve this by processing it in 'batches' of 1000. So it still checks the entire database, but in chunks of a known size. Each 'batch' would check the next 1000 and add the duplicates to a temporary table somewhere.

eileenmcnaughton commented:
I'm not quite sure how the batches work - let's say we have a DB of 300,000 or a group of 300,000 - ie. 30 * a batch of 1000. And lets assume it takes about 20 seconds to work through a batch of 1000 (or about 20 minutes to do the whole lot as a single batch since that is much slower).

So after 20 seconds we can calculated a group of matches into our prevnext_cache table. It is all the contacts that match the first 1000 contacts and could be anywhere from 10 to 30,000 contacts. What do we do now? Present people with these contacts or keep building our prevnext_cache? If the former then how do we explain what they are being presented with?

If the latter then they have a long wait just to see if there are any matches and while we protect the server we are still providing an unworkable interface on a large data set.

colemanw commented:
@JKingsnorth I saw your endorsement of this PR on a couple of issues. Could you fill in a bit more details about how you tested it? (adding a comment here on the PR is fine for that)

eileenmcnaughton commented:
(we could keep grabbing more by ajax in the back ground but that makes perfect sense if people are going to look at them & less sense if they are just going to click 'batch merge')

eileenmcnaughton commented:
@JKingsnorth I update the PR to include a setting as to the default limit. I realised that regardless of how we agreed to do the UI we had a bit of an impass on the default behaviour that could only be really resolved by a setting - so I added the setting rather than focus on fixing the rest of it. 

For most users the setting means there will be no change - but for large sites that are stupidly attached to their server uptime they can configure the setting & get the limit option to show up. This kind of means we can punt resolving the UI until we have time without missing out on the fixes in here.

I'm probably out of time to fix the flip critical issue for this release but there is one other critical in here - along with a fix I suspect is a partial fix on a data-loss critical issue - ie this is something I fixed 'in passing'

https://github.com/civicrm/civicrm-core/pull/8456/files#diff-874dc48dd188118ab0ba9cb9bbde9a02R867

yashodha commented:
Merging as per confirmation from @JoeMurray

JKingsnorth commented:
@colemanw Sorry, I should have been more specific. I followed the manual tests for the dedupe proceure which we've been putting together here: https://wiki.civicrm.org/confluence/pages/viewpage.action?pageId=356843540

JKingsnorth commented:
@JoeMurray @eileenmcnaughton  I still think the lack of an offset option effectively renders the 'limit' useless. If you specify a limit it's impossible to check the 'last' people in your database without checking the _whole_ database - which is what we are trying to avoid. I've created a follow-up issue here: https://issues.civicrm.org/jira/browse/CRM-19030

eileenmcnaughton commented:
Yeah - I would have done an offset but both you & Joe were committed to it defaulting to the whole DB so I figured I would have to do the setting anyway - which made the other much lower priority

JoeMurray commented:
@JKingsnorth agree that you need both offset and limit to make either worthwhile. My main concern is that the _default_ should be whole database, not that we shouldn't have option for limit etc

eileenmcnaughton commented:
Yep - it was because of the requirement for the default that I realised we needed a setting to control it - because the goal was to provide some protection by default. But, I didn't have time to work on both
